File: gitdot-web/packages/api/src/endpoint/migration/list-migrations.ts
import { z } from "zod";
import { MigrationResource, page } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListMigrationsRequest = z.object({
  cursor: z.string().optional(),
  limit: z.number().int().positive().optional(),
});
export type ListMigrationsRequest = z.infer<typeof ListMigrationsRequest>;

export const ListMigrationsResponse = page(MigrationResource);
export type ListMigrationsResponse = z.infer<typeof ListMigrationsResponse>;

export const ListMigrations = {
  path: "/migrations",
  method: "GET",
  request: ListMigrationsRequest,
  response: ListMigrationsResponse,
} as const satisfies Endpoint;
export type ListMigrations = typeof ListMigrations;

File: gitdot-api/src/endpoint/migration/list_migrations.rs
use serde::{Deserialize, Serialize};

use crate::{
    endpoint::Endpoint,
    resource::{common::Page, migration::MigrationResource},
};

pub struct ListMigrations;

impl Endpoint for ListMigrations {
    const PATH: &'static str = "/migrations";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListMigrationsRequest;
    type Response = ListMigrationsResponse;
}

#[derive(ApiRequest, Debug, Default, Serialize, Deserialize)]
pub struct ListMigrationsRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

pub type ListMigrationsResponse = Page<MigrationResource>;
